### PR TITLE
drop support for python3.5

### PR DESCRIPTION
**Pull request by graingert**
## Scope and purpose

Python 3.5 is EOL, this removes non-historical references to Python 3.5 and upgrades minimum python versions to py3.6 for various tools

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/9958
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] I have updated the automated tests.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.

### CODE REVIEW

**Comment by mthuurne** [comment]
There is already #1386 for dropping Python 3.5 support.

**Comment by graingert** [comment]
@mthuurne thanks, I've merged those changes in too

**Comment by rodrigc** [comment]
This is a duplicate of https://github.com/twisted/twisted/pull/1386

Please close it

**Review by rodrigc — DISMISSED** [review]
This is a duplicate.  Please close it.

**Comment by graingert** [comment]
> This is a duplicate of #1386
> 
> Please close it

Thanks, done

**Comment by rodrigc** [comment]
@graingert I didn't ask you to close and delete #1386 .  I asked you to close and delete #1528.

Please undo what you have done.

I view what you have done as aggressive and unwelcome behavior.

@glyph @exarkun @adiroiban

**Comment by graingert** [comment]
I assumed you wanted me to close #1386

**Comment by rodrigc** [comment]
You assumed wrong.  Please undo what you have done.

**Comment by graingert** [comment]
> You assumed wrong. Please undo what you have done.

Apologies, I've reopened #1386

**Comment by rodrigc** [comment]
Thank you.  Now please close #1528

**Comment by adiroiban** [comment]
@graingert I am sorry for the behavior displayed by Craig here.

I don't see any malice in your actions.

I do feel that Craig is bossy here and he is not following the community code of conduct.

I see the "Please" in the messages but they don't sound friendly at all.

The message that I see as offensive .

```
@graingert I didn't ask you to close and delete #1386 . I asked you to close and delete #1528.

Please undo what you have done.

I view what you have done as aggressive and unwelcome behavior.
```

**Comment by adiroiban** [comment]
> Thank you. Now please close #1528

I think that #1528 should be closed only if it has any critical technical issues.... like #1386 already solving all the issue and being merged.

This is open source so anyone should be free to work in parallel to any ticket.

I agree that working in parallel in the same issue is not ideal, but I don't see how it breaks the code of conduct.

Thomas commented on the  #1386 that  #1386 still has some py3.5 references and that he addressed those issues in this PR.

**Comment by adiroiban** [comment]
Also, I see that in the Trac ticket, https://twistedmatrix.com/trac/ticket/9958 the previous PR was not mentioned.

Thomas was the first one to add a link to a PR.

Again, In my view, Craig's behaviour is aggressive and not polite and intimidating.

**Comment by rodrigc** [comment]
@adiroiban @graingert I apologize if my behavior and communication in this PR was considered rude or aggressive by either of you.
That was not my intent.

Let me explain.

After working on the Twisted release on Saturday, I wanted to take a break from Twisted work on Sunday.

So I took a walk in the park.  While in the park, I took my phone with me, and was casually checking my e-mails while
walking in the park.

While walking in the park, I noticed that @graingert had submitted a patch to drop Python 3.5 support.
I already had a patch, approved by @wsanchez which I had worked on in September 2020.

Since @graingert is very fast at submitting PR's getting quick approvals, and getting them merged, I felt it was necessary
for me to respond, and indicate that I had already done work on this in #1386 .
I did not want @graingert to disregard the work that I had done in #1386, since I already did the initial work.  The only reason why I had not merged it was because I was waiting for a new Twisted release (last Twisted release to support 3.5) to come out.

However, since I was on my phone, it was very difficult for me to type full sentences and fully review #1528, which had a lot of commits in it.

So my sentences were a bit terse, and unfortunately may have come across as rude.
For that, I sincerely apologize.  That was not my intent.

I am not sure why @graingert interpreted my comments as instructions to delete my branch `9958-rodrigc-twisted-no-py3.5`.
This was unfortunate.  I admit, this did not leave me with good feelings.  I was still in the park at this time,
and was hoping that @graingert would just leave this PR alone for a while, so that I could focus on relaxing during the afternoon.

The good thing is that @graingert undid his deletion and restored my branch.  So that is good.

I apologize for any hurt feelings and misunderstandings that occurred during this PR.

@mthuurne I also apologize to you, if anything I said made you feel uncomfortable, since you commented in this PR.
I apologize if there

**Comment by rodrigc** [comment]
@adiroiban You are correct that in Trac ticket 9958 I did not originally put a reference to #1386 .

This was an honest mistake on my part, and I did not mean to do this.

I originally opened Trac ticket 9958, and meant to put a reference to #1386 in it.

I apologize to you and @graingert for my mistake, and for any misunderstanding that this caused.

**Comment by adiroiban** [comment]
> I apologize to you and @graingert for my mistake, and for any misunderstanding that this caused.

Accepted. The damage is done and at this moment I lost (for the second time) any hope in working together with you.

> Since @graingert is very fast at submitting PR's getting quick approvals, and getting them merged, I felt it was necessary
> for me to respond, and indicate that I had already done work on this in #1386 .
> I did not want @graingert to disregard the work that I had done in #1386, since I already did the initial work. The only reason why I had not merged it was because I was waiting for a new Twisted release (last Twisted release to support 3.5) to come out.

The work on Twisted is voluntary and done in free time. Nobody should tell on what anyone else should work.
Anyone should be free to create as many duplicate PR as wanted... and is up to the team to close them if they don't have any technical merit. 
You can inform that there is a duplicate PR but not ask a PR to be closed just because there is another unmerged / not approved yet PR.

Twisted development is slow enough... so I don't see why are you upset that the PRs made by Thomas are merged quickly.
I would say that have a PR quickly merged  is a good thing.

------------

I still think that the only way to improve collaboration is for Twisted dev to join IRC.
Over IRC I commented that I will wait before approving that PR and Marteen so that activity and commented that there is a duplicate PR.

https://freenode.logbot.info/twisted-dev/20210228#c7076608

Now, with your "release manager" role you sound like the boss who only stays in his office and never goes down in the workplace to mingle with the workers.

IRC is for quick collaboration and also for socialization.

**Comment by adiroiban** [comment]
image

I think that this is a sign that this PR is too big

**Review by adiroiban — CHANGES_REQUESTED** [review]
Thanks for working on this.

So... I am rejecting this as it's too much to check.

I see see that it now targes trunk.
When I did the review a few minutes ago, that was not the case.

I think that any PR should target trunk.
Use long lived feature branch only on exceptional occasions ... but I don't think that a
python3.5 removal and py3.6 syntax upgrade code required a long lived branch.

----


My suggested plan:

A. In a PR removed any 3.5 reference and do manual 3.5 changes
B. In another PR do the machine generated upgrade code: format syntax and type hints. This PR can be merged without a review of the code as soon as CI is green.
C. Maybe a quick PR in which we remove 3.5 from CI matrix...just to speed up any future work... but I am ok to do it with A.

-------

If anyone else is willing to review this PR as it is, I am fine to drop my review and have it merged.

**Comment by graingert** [comment]
@adiroiban I've removed the automated commits, and made a ticket for that https://twistedmatrix.com/trac/ticket/10108#comment:1

File: .github/workflows/test.yaml
Comment (adiroiban) [inline .github/workflows/test.yaml:22]: Maybe add a comment for others to understand what we are doing here:
```suggestion
          # Run on the minimum micro Python version that we can get on CI.
          # When updating the minimum Python version here, also update the
          # `python_requires` from `setup.cfg`.
          # Run on latest minor release of each major python version.
          python-version: ["3.6.7", 3.6, 3.7, 3.8, 3.9, pypy-3.6, pypy-3.7]
```

File: INSTALL.rst
Comment (adiroiban) [inline INSTALL.rst:9]: Maybe have it like this, so that we don't have to update this file  each time we support or remove a Python version.

```suggestion
- Python (see setup.cfg for supported versions)
```

File: docs/core/development/policy/coding-standard.rst
Comment (adiroiban) [inline docs/core/development/policy/coding-standard.rst]: Maybe 

```suggestion
Twisted is being ported to Python 3, see `setup.cfg` for the supported version.
```

File: docs/core/development/policy/coding-standard.rst
Comment (adiroiban) [inline docs/core/development/policy/coding-standard.rst]: Not sure if this paragraph has any current value.
It was created while we were still supporting 2.7.

I think that now, is common sense that any code should work on the minimum supported Python 3 version.

So, either remove this paragraph or replace it with something like:

```suggestion
All new modules must be compatible with the minimum supported Python version.

```

File: docs/core/development/policy/coding-standard.rst
Comment (adiroiban) [inline docs/core/development/policy/coding-standard.rst]: Again... maybe just remove this info... or write it in a generic way

```suggestion
New code in non-ported modules must be written to work with the minimum supported Python version (explicit bytes/unicode strings, new exception raising format, etc) as to prevent extra work when that module is eventually ported.
```

File: docs/core/development/policy/coding-standard.rst
Comment (adiroiban) [inline docs/core/development/policy/coding-standard.rst]: I have no idea which MSVC++ version is required ...I remember that the C++ compiler has now a different name.. .like Windows 10 SDK or something like this.

But I think that the C code poilicy is now to move any C code outside of this repo :)

```suggestion
C code must be optional, and work across multiple platforms (MSVC++9/10/14 for Python 3 on Windows, as well as recent GCCs and Clangs for Linux, macOS, and FreeBSD).

It should be distributed as a separate package and not  as part of the main Twisted project.
```

File: docs/core/howto/python3.rst
Comment (adiroiban) [inline docs/core/howto/python3.rst:7]: Not sure if we should keep this line... I guess that by this time, someone will already know the supported Python3 version for Twisted

File: setup.cfg
Comment (adiroiban) [inline setup.cfg]: Just to be extra sure and match what we use for CI :)

```suggestion
# When updating this value, make sure our CI matrix include the minimum version.
python_requires = >=3.6.7
```

**Review by adiroiban — APPROVED** [review]
Looks good. Many thanks! Only minor comments.

File: docs/core/development/policy/coding-standard.rst
Comment (graingert) [reply docs/core/development/policy/coding-standard.rst]: how about?

```suggestion
Twisted was ported to Python 3.
```

File: docs/core/development/policy/coding-standard.rst
Comment (graingert) [reply docs/core/development/policy/coding-standard.rst]: yeah I think this being removed makes the most sense, CI already enforced the relevant python versions

```suggestion
```

File: docs/core/development/policy/coding-standard.rst
Comment (graingert) [reply docs/core/development/policy/coding-standard.rst]: > extra work when that module is eventually ported.

yep, we don't have a concept of ported/non-ported modules anymore
```suggestion
```

File: docs/core/development/policy/coding-standard.rst
Comment (graingert) [reply docs/core/development/policy/coding-standard.rst]: this policy is already here:
https://github.com/twisted/twisted/pull/1528/files/92cd5ad286fc947645867667f218b12eeb9459ec#diff-514cad77343003f66671092af193f29ac569e2a6b6bd04ffd62ff468f17b6f07R644

Comment (graingert) [reply docs/core/development/policy/coding-standard.rst]:
```suggestion
C code must be optional, and work across multiple platforms (MSVC++14 for Python 3 on Windows, as well as recent GCCs and Clangs for Linux, macOS, and FreeBSD).
```

https://wiki.python.org/moin/WindowsCompilers#Which_Microsoft_Visual_C.2B-.2B-_compiler_to_use_with_a_specific_Python_version_.3F

File: setup.cfg
Comment (graingert) [reply setup.cfg]:
```suggestion
# When updating this value, make sure our CI matrix includes a matching minimum version.
python_requires = >=3.6.7
```

File: docs/core/howto/python3.rst
Comment (graingert) [reply docs/core/howto/python3.rst:7]: I'm going to err on the side of keeping it for now

File: INSTALL.rst
Comment (graingert) [reply INSTALL.rst:9]: I don't think it's too much of a burden to update it everywhere. It's only once per year or so

**Review by adiroiban — APPROVED** [review]
Assuming pypy tests will be over soon, this is all ok and can be merged. Thanks!

File: src/twisted/__init__.py
Comment (wsanchez) [inline src/twisted/__init__.py]: I propose that we remove this check.  It's uncovered, so blocking this PR, and `setup.py` does the job already.

**Comment by wsanchez** [comment]
FYI I've hidden the process back-and-forth, which I think we've agreed to move on from, so we can focus on the changes here.

**Review by wsanchez — APPROVED** [review]

**Pull request merged**
